File: TS/系统/02．物品系统/15．装备技能/00．物品/21．使者精神魔杖.ts
/** @noSelfInFile */

const jass = require("jass.common") as any;

const { YDUserDataGet, YDUserDataSet, YDUserDataHas, YDUserDataClear } = require("lib.扩展函数.YDWE函数.index") as {
  YDUserDataGet: (this: void, tableType: string, tableKey: any, attr: string, valueType: string) => any;
  YDUserDataSet: (this: void, tableType: string, tableKey: any, attr: string, valueType: string, value: any) => void;
  YDUserDataHas: (this: void, tableType: string, tableKey: any, attr: string, valueType: string) => boolean;
  YDUserDataClear: (this: void, tableType: string, tableKey: any, attr: string, valueType: string) => void;
};
const { stringToFourCCSafe } = require("lib.扩展函数.封装函数.01．通用工具.01．FourCC转换安全版") as {
  stringToFourCCSafe: (this: void, s: string | undefined | null) => number;
};

const GetItemTypeId = jass.GetItemTypeId as (item: any) => number;
const GetUnitX = jass.GetUnitX as (unit: any) => number;
const GetUnitY = jass.GetUnitY as (unit: any) => number;
const GetUnitTypeId = jass.GetUnitTypeId as (unit: any) => number;
const GetUnitFacing = jass.GetUnitFacing as (unit: any) => number;
const GetOwningPlayer = jass.GetOwningPlayer as (unit: any) => any;
const IsUnitRace = jass.IsUnitRace as (unit: any, race: any) => boolean;
const IsHeroUnitId = jass.IsHeroUnitId as (unitId: number) => boolean;
const KillUnit = jass.KillUnit as (unit: any) => void;
const CreateUnit = jass.CreateUnit as (player: any, unitId: number, x: number, y: number, facing: number) => any;
const UnitApplyTimedLife = jass.UnitApplyTimedLife as (unit: any, buffId: number, duration: number) => void;
const CreateTimer = jass.CreateTimer as () => any;
const TimerStart = jass.TimerStart as (timer: any, timeout: number, periodic: boolean, callback: (this: void) => void) => void;
const GetExpiredTimer = jass.GetExpiredTimer as () => any;
const GetHandleId = jass.GetHandleId as (handle: any) => number;
const DestroyTimer = jass.DestroyTimer as (timer: any) => void;
const RACE_DEMON = jass.RACE_DEMON as any;

import type { 物品技能事件上下文 } from "../03．主动技能/03．物品使用触发/01．物品使用触发常量";
import { 使者精神魔杖物品ID } from "../03．主动技能/00．公共/01．主动技能物品ID";
import { 使者精神魔杖配置 } from "../03．主动技能/03．物品使用触发/00．物品使用触发配置";

const 清理表: Record<number, any | undefined> = {};
const 限时生命BuffID = stringToFourCCSafe("BHwe");

function 是否为使者精神魔杖(this: void, 物品: any): boolean {
  if (物品 == null || 物品 === 0) return false;
  return GetItemTypeId(物品) === 使者精神魔杖物品ID;
}

function 目标可存储(this: void, 目标单位: any): boolean {
  if (目标单位 == null || 目标单位 === 0) return false;
  if (IsUnitRace(目标单位, RACE_DEMON)) return false;
  return !IsHeroUnitId(GetUnitTypeId(目标单位));
}

function on使者精神魔杖存储过期(this: void): void {
  const timer = GetExpiredTimer();
  const timerID = GetHandleId(timer);
  const 施法单位 = 清理表[timerID];
  delete 清理表[timerID];
  if (施法单位 != null && 施法单位 !== 0) {
    YDUserDataClear("unit", 施法单位, 使者精神魔杖配置.存储字段, "unitcode");
  }
  DestroyTimer(timer);
}

function 启动存储过期计时(this: void, 施法单位: any): void {
  const timer = CreateTimer();
  if (timer == null || timer === 0) return;
  清理表[GetHandleId(timer)] = 施法单位;
  TimerStart(timer, 使者精神魔杖配置.存储持续时间, false, on使者精神魔杖存储过期);
}

export function 处理使者精神魔杖使用(this: void, 上下文: 物品技能事件上下文): void {
  if (!是否为使者精神魔杖(上下文.物品)) return;
  const 施法单位 = 上下文.施法单位;
  if (施法单位 == null || 施法单位 === 0) return;

  const 已存储 = YDUserDataHas("unit", 施法单位, 使者精神魔杖配置.存储字段, "unitcode");
  const 目标单位 = 上下文.目标单位;
  if (!已存储) {
    if (!目标可存储(目标单位)) return;
    KillUnit(目标单位);
    YDUserDataSet("unit", 施法单位, 使者精神魔杖配置.存储字段, "unitcode", GetUnitTypeId(目标单位));
    启动存储过期计时(施法单位);
    return;
  }

  const 存储单位类型 = YDUserDataGet("unit", 施法单位, 使者精神魔杖配置.存储字段, "unitcode");
  const x = 目标单位 == null || 目标单位 === 0 ? 上下文.目标X : GetUnitX(目标单位);
  const y = 目标单位 == null || 目标单位 === 0 ? 上下文.目标Y : GetUnitY(目标单位);
  const 召唤单位 = CreateUnit(GetOwningPlayer(施法单位), 存储单位类型, x, y, GetUnitFacing(施法单位));
  if (召唤单位 != null && 召唤单位 !== 0) {
    UnitApplyTimedLife(召唤单位, 限时生命BuffID, 使者精神魔杖配置.召唤持续时间);
  }
}

export {};

File: src/系统/02．物品系统/15．装备技能/00．物品/21．使者精神魔杖.lua
local ____lualib = require("lualib_bundle")
local __TS__Delete = ____lualib.__TS__Delete
local ____exports = {}
local ____01_FF0E_4E3B_52A8_6280_80FD_7269_54C1ID = require("系统.02．物品系统.15．装备技能.03．主动技能.00．公共.01．主动技能物品ID")
local _____4F7F_8005_7CBE_795E_9B54_6756_7269_54C1ID = ____01_FF0E_4E3B_52A8_6280_80FD_7269_54C1ID["使者精神魔杖物品ID"]
local ____00_FF0E_7269_54C1_4F7F_7528_89E6_53D1_914D_7F6E = require("系统.02．物品系统.15．装备技能.03．主动技能.03．物品使用触发.00．物品使用触发配置")
local _____4F7F_8005_7CBE_795E_9B54_6756_914D_7F6E = ____00_FF0E_7269_54C1_4F7F_7528_89E6_53D1_914D_7F6E["使者精神魔杖配置"]
local ____require_result_0 = require("lib.扩展函数.自定义扩展函数.03．调试输出")
local debugLogForce = ____require_result_0.debugLogForce
local jass = require("jass.common")
local ____require_result_1 = require("lib.扩展函数.YDWE函数.index")
local YDUserDataGet = ____require_result_1.YDUserDataGet
local YDUserDataSet = ____require_result_1.YDUserDataSet
local YDUserDataHas = ____require_result_1.YDUserDataHas
local YDUserDataClear = ____require_result_1.YDUserDataClear
local ____require_result_2 = require("lib.扩展函数.封装函数.01．通用工具.01．FourCC转换安全版")
local stringToFourCCSafe = ____require_result_2.stringToFourCCSafe
local GetItemTypeId = jass.GetItemTypeId
local GetUnitX = jass.GetUnitX
local GetUnitY = jass.GetUnitY
local GetUnitTypeId = jass.GetUnitTypeId
local GetUnitFacing = jass.GetUnitFacing
local GetOwningPlayer = jass.GetOwningPlayer
local IsUnitRace = jass.IsUnitRace
local IsHeroUnitId = jass.IsHeroUnitId
local KillUnit = jass.KillUnit
local CreateUnit = jass.CreateUnit
local UnitApplyTimedLife = jass.UnitApplyTimedLife
local CreateTimer = jass.CreateTimer
local TimerStart = jass.TimerStart
local GetExpiredTimer = jass.GetExpiredTimer
local GetHandleId = jass.GetHandleId
local DestroyTimer = jass.DestroyTimer
local RACE_DEMON = jass.RACE_DEMON
local _____6E05_7406_8868 = {}
local _____9650_65F6_751F_547DBuffID = stringToFourCCSafe("BHwe")
local function _____662F_5426_4E3A_4F7F_8005_7CBE_795E_9B54_6756(_____7269_54C1)
    if _____7269_54C1 == nil or _____7269_54C1 == 0 then
        return false
    end
    return GetItemTypeId(_____7269_54C1) == _____4F7F_8005_7CBE_795E_9B54_6756_7269_54C1ID
end
local function _____76EE_6807_53EF_5B58_50A8(_____76EE_6807_5355_4F4D)
    if _____76EE_6807_5355_4F4D == nil or _____76EE_6807_5355_4F4D == 0 then
        return false
    end
    if IsUnitRace(_____76EE_6807_5355_4F4D, RACE_DEMON) then
        return false
    end
    return not IsHeroUnitId(GetUnitTypeId(_____76EE_6807_5355_4F4D))
end
local function ____on_4F7F_8005_7CBE_795E_9B54_6756_5B58_50A8_8FC7_671F()
    local timer = GetExpiredTimer()
    local timerID = GetHandleId(timer)
    local _____65BD_6CD5_5355_4F4D = _____6E05_7406_8868[timerID]
    __TS__Delete(_____6E05_7406_8868, timerID)
    if _____65BD_6CD5_5355_4F4D ~= nil and _____65BD_6CD5_5355_4F4D ~= 0 then
        YDUserDataClear("unit", _____65BD_6CD5_5355_4F4D, _____4F7F_8005_7CBE_795E_9B54_6756_914D_7F6E["存储字段"], "unitcode")
    end
    DestroyTimer(timer)
end
local function _____542F_52A8_5B58_50A8_8FC7_671F_8BA1_65F6(_____65BD_6CD5_5355_4F4D)
    local timer = CreateTimer()
    if timer == nil or timer == 0 then
        return
    end
    _____6E05_7406_8868[GetHandleId(timer)] = _____65BD_6CD5_5355_4F4D
    TimerStart(timer, _____4F7F_8005_7CBE_795E_9B54_6756_914D_7F6E["存储持续时间"], false, ____on_4F7F_8005_7CBE_795E_9B54_6756_5B58_50A8_8FC7_671F)
end
____exports["处理使者精神魔杖使用"] = function(_____4E0A_4E0B_6587)
    debugLogForce("21．使者精神魔杖", "进入", "处理使者精神魔杖使用")
    if not _____662F_5426_4E3A_4F7F_8005_7CBE_795E_9B54_6756(_____4E0A_4E0B_6587["物品"]) then
        return
    end
    local _____65BD_6CD5_5355_4F4D = _____4E0A_4E0B_6587["施法单位"]
    if _____65BD_6CD5_5355_4F4D == nil or _____65BD_6CD5_5355_4F4D == 0 then
        return
    end
    local _____5DF2_5B58_50A8 = YDUserDataHas("unit", _____65BD_6CD5_5355_4F4D, _____4F7F_8005_7CBE_795E_9B54_6756_914D_7F6E["存储字段"], "unitcode")
    local _____76EE_6807_5355_4F4D = _____4E0A_4E0B_6587["目标单位"]
    if not _____5DF2_5B58_50A8 then
        if not _____76EE_6807_53EF_5B58_50A8(_____76EE_6807_5355_4F4D) then
            return
        end
        KillUnit(_____76EE_6807_5355_4F4D)
        YDUserDataSet(
            "unit",
            _____65BD_6CD5_5355_4F4D,
            _____4F7F_8005_7CBE_795E_9B54_6756_914D_7F6E["存储字段"],
            "unitcode",
            GetUnitTypeId(_____76EE_6807_5355_4F4D)
        )
        _____542F_52A8_5B58_50A8_8FC7_671F_8BA1_65F6(_____65BD_6CD5_5355_4F4D)
        return
    end
    local _____5B58_50A8_5355_4F4D_7C7B_578B = YDUserDataGet("unit", _____65BD_6CD5_5355_4F4D, _____4F7F_8005_7CBE_795E_9B54_6756_914D_7F6E["存储字段"], "unitcode")
    local x = (_____76EE_6807_5355_4F4D == nil or _____76EE_6807_5355_4F4D == 0) and _____4E0A_4E0B_6587["目标X"] or GetUnitX(_____76EE_6807_5355_4F4D)
    local y = (_____76EE_6807_5355_4F4D == nil or _____76EE_6807_5355_4F4D == 0) and _____4E0A_4E0B_6587["目标Y"] or GetUnitY(_____76EE_6807_5355_4F4D)
    local _____53EC_5524_5355_4F4D = CreateUnit(
        GetOwningPlayer(_____65BD_6CD5_5355_4F4D),
        _____5B58_50A8_5355_4F4D_7C7B_578B,
        x,
        y,
        GetUnitFacing(_____65BD_6CD5_5355_4F4D)
    )
    if _____53EC_5524_5355_4F4D ~= nil and _____53EC_5524_5355_4F4D ~= 0 then
        UnitApplyTimedLife(_____53EC_5524_5355_4F4D, _____9650_65F6_751F_547DBuffID, _____4F7F_8005_7CBE_795E_9B54_6756_914D_7F6E["召唤持续时间"])
    end
end
return ____exports
